packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts

```diff
@@ -0,0 +1,70 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+
+export const messageId = 'noUnnecessaryParameterPropertyAssignment';
+
+export default util.createRule({
+  name: 'no-unnecessary-parameter-property-assignment',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Disallow unnecessary assignment of constructor property parameter to itself',
+      recommended: 'error',
+      requiresTypeChecking: false,
+    },
+    schema: [],
+    messages: {
+      [messageId]:
+        'Assignment has no effect as roperty parameter is already defined in the constructor.',
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    const parameterProperties = new Set<string>();
+    return {
+      'ClassDeclaration, ClassExpression'(): void {
+        parameterProperties.clear();
+      },
+      'MethodDefinition[kind="constructor"] > FunctionExpression > TSParameterProperty'(
+        property: TSESTree.TSParameterProperty,
+      ): void {
+        if (property.parameter.type !== AST_NODE_TYPES.Identifier) {
+          return;
```

Comment:
[Bug] Hmm, I think this is a case we'll want to tackle. Why limit to just identifiers?

```ts
class A {
  constructor(b: 'c' | 'd') {
    this.b = b;
  }
}
```

Suggestion: remove this `if` and the `Identifier` check later?